Adding custom module to find pid of the process

saranyasridharan opened the pull request:
SUMMARY
This module will find the pid of the given process name either locally or in remote system, depending on the host parameter.


 ISSUE TYPE

 - New Module Pull Request


COMPONENT NAME
pids

ANSIBLE VERSION
ansible 2.8.0.dev0  last updated 2018/08/07 12:03:58 (GMT +550)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/ansible/newcode/library']
  ansible python module location = /home/ansible/lib/ansible
  executable location = /home/ansible/bin/ansible
  python version = 2.7.5 (default, Jul 13 2018, 13:06:57) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]


 ADDITIONAL INFORMATION
I have used pidof module to find the process IDs(PIDs) of the given process

jamescassell reviewed (commented):
Looks like it will find processes that match a substring. If I look for 'id', it will find 'squid'

saranyasridharan commented:
Hi James,

        Shall I change it to exact string  matching?

Thanks,
Saranya

On Thu, Aug 9, 2018 at 5:33 AM James Cassell <notifications@github.com>
wrote:

> *@jamescassell* commented on this pull request.
>
> Looks like it will find processes that match a substring. If I look for
> 'id', it will find 'squid'
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/43807#pullrequestreview-144651582>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APAAqy-KI3bbwsHXksjV9Vpt_aJ5il0fks5uO3xigaJpZM4VzSdp>
> .
>

saranyasridharan commented:
@jamescassell Changed as per your review command, Now it will check only whole process name instead of substring

felixfontein commented on lib/ansible/modules/system/pid.py:
```diff
@@ -0,0 +1,55 @@
+pid:
+  description: It return process id of the given process
+  returned: list of process id
+  type: dict
```
Is it a `list` or a `dict`? I think it is a `list`.

felixfontein commented on lib/ansible/modules/system/pid.py:
```diff
@@ -0,0 +1,55 @@
+
+def get_pid(name):
+    processid = subprocess.check_output("pidof %s" % (name), shell=True)
+    return processid.split("\n")
```
On my machine (Linux), the output of `pidof` is separated by spaces, not by newlines.

Also, when `name` contains spaces, this doesn't work. It's better to pass a list of strings like this: `["pidof", name]`, instead of a precomposed string.

gundalow commented:
I wonder if this should be a `lookup` plugin, rather than a module https://docs.ansible.com/ansible/devel/plugins/lookup.html & https://docs.ansible.com/ansible/devel/dev_guide/developing_plugins.html#lookup-plugins

dagwieers commented:
@gundalow A lookup plugin can *only* run on the Ansible host, whereas a module can run on the host or on target nodes.

saranyasridharan commented:
@felixfontein I have addressed your comments.
  - Changed return type 
  -  Added split using space
  -  command is given in the form of list ["pidof", name], instead of a precomposed string.
Thanks for the comments. Please look into that and let me know if anything else is needed.

felixfontein commented on lib/ansible/modules/system/pid.py:
```diff
@@ -0,0 +1,56 @@
+DOCUMENTATION = '''
+module: pid
+version_added: 2.8
+description: "Gives the pid of the given process name"
```
This should mention that the result is a list of none, one, or more PIDs. Also, please mention both "process ID" and "PID" so that readers who know only one of those terms understand that they mean the same.

felixfontein commented on lib/ansible/modules/system/pid.py:
```diff
@@ -0,0 +1,56 @@
+author:
+  - Saranya Sridharan
+requirements:
+  - only standard library needed
```
That's not true. You need an external program called `pidof`. This is *not* installed on every Unix (see http://macappstore.org/pidof/).

felixfontein commented on lib/ansible/modules/system/pid.py:
```diff
@@ -0,0 +1,56 @@
+  name:
+    description: the name of the process you want to get pid for
+    required: true
+    default: null
```
Don't specify default values for required options. Simply remove this line.

felixfontein commented on lib/ansible/modules/system/pid.py:
```diff
@@ -0,0 +1,56 @@
+# Pass the process name
+- name: Getting pid of the process
+  pid:
+      name: python
```
You should add `register: pid_of_python` and then a second task which uses it (like `- debug: msg="PIDs of Python: {{ ', '.join(pid_of_python.pid) }}"`).

felixfontein commented on lib/ansible/modules/system/pid.py:
```diff
@@ -0,0 +1,56 @@
+      name: python
+'''
+RETURN = '''
+pid:
```
Why `pid` and not `pids`, since this is actually a list of PIDs?

felixfontein commented on lib/ansible/modules/system/pid.py:
```diff
@@ -0,0 +1,56 @@
+def get_pid(name):
+    processid = subprocess.check_output(["pidof", name])
+    processid = processid.strip("\n")
+    return processid.split(" ")
```
Why not convert the pieces of the splitted string to `int`s? Either here or in line 51?

felixfontein commented on test/integration/targets/pid/tasks/main.yml:
```diff
@@ -0,0 +1,17 @@
+  assert:
+    that:
+    - emptypids is not changed
+    - emptypids.pid == [""] 
```
If the process isn't running, the result should not be a list containing an empty string, but an empty list!

felixfontein reviewed (changes requested):
Some more ideas, and some more changes :)

saranyasridharan commented:
@felixfontein I have made changed as you mentioned, But problem occured while committing. Please help me with this.

felixfontein commented:
@saranyasridharan it looks like something went wrong. Did you do a merge with `devel`? Best is probably to update your `devel` branch to upstream (see [here](https://docs.ansible.com/ansible/devel/dev_guide/developing_rebasing.html) for details), and then do a `git rebase -i devel` and remove everything which isn't by you (or which is a merge commit) from the list.

saranyasridharan committed:
Addressed changes mentioned in the comment

saranyasridharan committed:
Adding custom module to find PID of the process

saranyasridharan committed:
Adding custom module to find PID of the process

saranyasridharan commented:
I have tried the rebase. Then only this problem happened. Can you please  tell me how do I remove the merge commit and other files which is not commited by me ? I have tried git rebase -i commit number it didnt give any response for long time. @felixfontein

felixfontein commented:
Hmm, no idea. Maybe the easiest is to copy the files you added somewhere else (so they don't get lost), delete the `pid` branch locally, checkout the `devel` branch, re-create the `pid` branch locally (from `devel`) and re-add the files, and then force-push that branch to Github. Then you should have a new clean shiny branch with a single commit with the latest state of the module and its tests.

saranyasridharan committed:
Delete pids.py

I will work on my separate branch as of now

saranyasridharan committed:
Delete main.yml

I will work on my branch as of now

saranyasridharan committed:
Delete aliases

I will work on my branch as of now

saranyasridharan commented:
@felixfontein Thanks so much for the help. I have tried in fresh checkout it worked as you mentioned. I have addressed all the comments provided. Please say if anything else is needed.

felixfontein commented:
@saranyasridharan It looks much better now :) I don't have time today, but I'll look at it during the next days.

saranyasridharan commented:
Sure. Thanks @felixfontein.

On Mon, 24 Sep 2018 at 5:35 PM, Felix Fontein <notifications@github.com>
wrote:

> @saranyasridharan <https://github.com/saranyasridharan> It looks much
> better now :) I don't have time today, but I'll look at it during the next
> days.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/43807#issuecomment-423951964>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APAAq264pLLIL5fxJjfXQ_2IplAFsIuOks5ueMqmgaJpZM4VzSdp>
> .
>

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,60 @@
+module: pids
+version_added: 2.8
+description: "Gives either one or more pids/process ids  of the given process name or simply empty list if the process does not exist"
+short_description: "Gives either one or more pids/process ids  of the given process name or simply empty list if the process does not exist"
```
The short description should be shorter; it will be shown next to the module name as part of the title of the module's documentation. Example: https://docs.ansible.com/ansible/devel/modules/acme_certificate_module.html there, `short_description` is `Create SSL/TLS certificates with the ACME protocol`. `description` is shown below `Synopsis`.

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,60 @@
+- name: Getting pids/process ids of the process
+  pids:
+      name: python
+      register: pids_of_python
```
`register` should be one level up (at the same level as `pids:`)

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,60 @@
+      register: pids_of_python
+
+- name: Getting the pids/process ids obtained
+- debug: msg="PIDS of python:{{','.join(pids_of_python.pid)}}"
```
The `-` in front of `debug:` should not be there if `name:` belongs to `debug:`.

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,60 @@
+RETURN = '''
+pids:
+  description: It return process ids/pids of the given process
+  returned: list of empty,one/more process ids/pids
```
there's a space missing between `empty,` and `one/more`. I'd also write out `/` as ` or `, and replace `empty` by `none`.

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,60 @@
+
+RETURN = '''
+pids:
+  description: It return process ids/pids of the given process
```
There's no need to always write `process ids/pids`. Do it once (in the `description`) and from then on only use `PIDs` or `process IDs`, but not both.

Also, don't use `/`, instead write `process IDs (PIDs)`.

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,60 @@
+def main():
+    module = AnsibleModule(argument_spec={"name": {"required": True, "type": "str"}})
+    name = module.params["name"]
+    response = dict(pid=get_pid(name))
```
That should be `dict(pids=...`, as you wrote above that the return value is called `pids`.

felixfontein commented on test/integration/targets/pid/aliases:
```diff
@@ -0,0 +1 @@
+unsupported
```
When you really want to rename the module to `pids` (instead of `pid`), you should also rename the integration test target directory from `pid` to `pids`.

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,60 @@
+DOCUMENTATION = '''
+module: pids
+version_added: 2.8
+description: "Gives either one or more pids/process ids  of the given process name or simply empty list if the process does not exist"
```
I would phrase this differently. Maybe something like this: `Retrieves a list of process IDs (PIDs) of all processes of the givne name. Returns an empty list if no process of the iven name exists.`

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,60 @@
+description: "Gives either one or more pids/process ids  of the given process name or simply empty list if the process does not exist"
+short_description: "Gives either one or more pids/process ids  of the given process name or simply empty list if the process does not exist"
+author:
+  - Saranya Sridharan
```
You should use the format described [here](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#fields), i.e. `Saranya Sridharan (@saranyasridharan)`. This allows ansibot to automatically ping you when issues / PRs / ... involving this module are created.

felixfontein reviewed (changes requested):
Some more mostly minor changes :)

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,60 @@
+  - Needs 'pidof' to be installed in the system
+options:
+  name:
+    description: the name of the process you want to get pid for
```
I think you should either use `pid` or `PID` everywhere. Whether you prefer `PID` or `pid` is something you have to decide :)

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,60 @@
+RETURN = '''
+pids:
+  description: It return process IDs of the given process
+  returned: list of none, one or, more process IDs
```
the comma should be before `or`, i.e. `list of none, one, or more process IDs`

felixfontein commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,17 @@
+  assert:
+    that:
+    - emptypids is not changed
+    - emptypids.pid == [] 
```
Here you still use `pid` instead of `pids`.

felixfontein commented on test/integration/targets/pids/aliases:
```diff
@@ -0,0 +1 @@
+unsupported
```
Why can't this test be run in CI? It would be especially useful to see whether it works everywhere.

saranyasridharan replied:
How do I verify whether it runs in CI or not ?

felixfontein replied:
The easiest is to try it out :)

Simply put `shippable/posix/group3` in here instead of `unsupported`, and see what happens. Maybe you will get some test failures, but at least then you get an idea why they fail where (for example, `pidof` might not be available somewhere)

Maybe you should also add some more tests which probably match processes running on the machine (like check for `python`, `python2`, `python3`) to see whether something bad will happen if something is found.

felixfontein commented:
Now a lot of test fail, but that's a Good Thing(tm): it shows that `subprocess.check_output(["pidof", name])` in general doesn't work. You want to take a look at `module.get_bin_path('pidof', True)`, which tries to find the correct path where `pidof` resides for you.

Also, instead of using `subprocess` directly, it's probably better to use `module.run_command()`. See for example [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/acme.py#L212) on how to use it (or just grep for `module.run_command` in `module_utils` or `modules` and see a lot of more examples).

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,62 @@
+
+def get_pid(name,module):
+    command = module.get_bin_path('pidof', True)
+	command+ = name
```
You have a lot of spacing problems in the above three lines. Please install `flake8` on your machine (with `pip install flake8`) and run it for this module (`flake8 pids.py`), that will show you a lot of the errors much earlier.

Also, `command` is a string, not an array, so you either need to write `command = [command, name]`, or you do `module.run_command([command, name])` in the next line.

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/85506/54/tests

felixfontein commented:
@saranyasridharan If you inspect the failing tests, you'll see that there are different results. On some of the systems, `pidof` doesn't exist. On others, there's another error:
```
[...]
  File "/root/.ansible/tmp/ansible-tmp-1537863809.52-19689758896697/AnsiballZ_pids.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_pids_payload_mTj_B4/__main__.py", line 61, in <module>
  File "/tmp/ansible_pids_payload_mTj_B4/__main__.py", line 56, in main
  File "/tmp/ansible_pids_payload_mTj_B4/__main__.py", line 49, in get_pid
AttributeError: 'tuple' object has no attribute 'strip'
```
The reason is simple: `module.run_command()` doesn't simply return the stdout as a string. You should always check the documentation of something before using it. It says: `Execute a command, returns rc, stdout, and stderr.` So you need to write (to get only stdout) `dummy, processid, dummy = module.run_command([command, name])` (and then in the next line `processid = processid.strip('\n')`).

This should fix most of the failing tests.

For the other ones, `pidof` is probably not around, so you need to disable them. Add `skip/freebsd` and `skip/osx` to your `aliases` file. That will make sure that the tests aren't run on the FreeBSD and OSX targets (which fail because `pidof` is missing). This also shows that the requirement `Needs 'pidof' to be installed in the system` was necessary.

saranyasridharan commented:
> The current error is because if the stdout is the empty string (`''`), the split of it will be a list containing one empty string as an entry. You have to inspect the result after calling `split(" ")` and change it to `[]` if it only contains the empty string (or do something like `xxx.split(" ") if xxx else []`).

Oh Thanks. I will check this. I was waiting for mail from Ansibot. I couldn't get the error.

felixfontein commented:
Ansibot won't post the information about such errors. To see them, look for the "All tests have failed" at the bottom of the PR page (https://github.com/ansible/ansible/pull/43807) and open the "Details" link for shippable. There, you can scroll through the list of tests and click on one of the red (failed) ones. There, you can see the console output and more infos about what happened.

saranyasridharan commented:
> Ansibot won't post the information about such errors. To see them, look for the "All tests have failed" at the bottom of the PR page (#43807) and open the "Details" link for shippable. There, you can scroll through the list of tests and click on one of the red (failed) ones. There, you can see the console output and more infos about what happened.

sure thanks. Now I have checked, it is not finding the path of 'pidof' in centos7

felixfontein commented:
Yes, skip that test.

sivel commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,62 @@
+  register: pids_of_python
+
+- name: Printing the process IDs obtained
+  debug: msg="PIDS of python:{{','.join(pids_of_python.pid)}}"
```
This debug needs to be updated for a few things:

1. Use full YAML syntax, not `key=value`
1. `.pid` should be `.pids`
1. Instead of `','.join` please use the `|join` jinja filter.

saranyasridharan replied:
I have changed accordingly. thanks

sivel commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,62 @@
+    command = module.get_bin_path('pidof', True)
+    rc, processid, stderr = module.run_command([command, name])
+    processid = processid.strip("\n")
+    result = map(int, processid.split(" ")) if processid else []
```
For python3 compatibility, you will need to wrap `map` with `list` as python3 doesn't return a list.

It would probably just be best to use a list comprehension instead.

saranyasridharan replied:
I have changed accordingly. thanks

sivel commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,62 @@
+
+
+def main():
+    module = AnsibleModule(argument_spec={"name": {"required": True, "type": "str"}})
```
Please break this out onto multiple lines.  If additional arguments are added later or additional customizations are needed, this format will require more changes later.

saranyasridharan replied:
Done as you have suggested. Thanks

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,23 @@
+  assert:
+    that:
+    - emptypids is not changed
+    - emptypids.pids == [] 
```
Can you include a test that returns results as well? Ideally the test would cover cases of 0, 1 and 2 PIDs returned. Perhaps add a simple python script which sleeps for 10 seconds, run it, and use the module to find the PID(s).

saranyasridharan commented:
I have addressed all the review comments. But I have a problem. When I run 'pidof' in linux command prompt, it gives few pids , and when I ran the pidof in the ansible command , it is giving different pids, But it shouldnt be the case always right? Can anyone help me to find where am I missing ?

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,52 @@
+    mode: 0777
+    
+- name: Executing the python script
+  script: /home/sample.py
```
I think you'll need to change two things here:

1. Execute the script using [async](https://docs.ansible.com/ansible/devel/user_guide/playbooks_async.html) so that the next task runs before the script exits.
2. Use [setproctitle](https://pypi.org/project/setproctitle/) to set the process name of your script.

Without the 2nd change you won't be able to distinguish between your script and other python processes, including the ones running ansible-playbook and the module.

mattclay replied:
You'll also need to install `setproctitle` in the test so it's available for use.

saranyasridharan replied:
sure. I will install that through test. I couldn't find how to use setproctitle. Can you show me one example @mattclay ?

mattclay commented on test/integration/targets/pids/sample.py:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env python
+import time
```
```python
import setproctitle


setproctitle.setproctitle('some-unique-name-goes-here')

saranyasridharan commented:
@mattclay I am getting an error saying that pidof module is returning non zero return code, I thought may be the script is not running till I get the process id, So I have used more sleep time. Still it is failing with the same error. Do you know any idea why is it so ?

felixfontein commented:
`man pidof` says (among others):
```
EXIT STATUS
       0      At least one program was found with the requested name.
       1      No program was found with the requested name.
```
Now the command module by default fails when a non-zero return code is returned. You can change this behavior by adding `failed_when: no` to the task. (See https://docs.ansible.com/ansible/2.6/user_guide/playbooks_error_handling.html#controlling-what-defines-failure)

saranyasridharan commented:
@mattclay @felixfontein There seems to be some error , But I couldnt find any indentation error. Can you please help me with this ?

felixfontein commented:
Change
```
   - "{{pids['pids']|join(' ')}} == {{pidsobtained['stdout']}}"
```
to
```
   - "pids['pids']|join(' ') == pidsobtained['stdout']"
```
That should fix that error.

felixfontein commented:
(Also, you can write `pids.pids | join(' ') == pidsobtained.stdout` -- that's more readable.)

saranyasridharan commented:
> (Also, you can write `pids.pids | join(' ') == pidsobtained.stdout` -- that's more readable.)

Thanks so much that solved my issue :)

saranyasridharan commented:
@mattclay @felixfontein @sivel I have addressed your review comments as better as I could. Please review this PR and let me know if anything else is needed.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,59 @@
+ 
+
+
+  
```
Remove extra blank lines.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,59 @@
+    - emptypids.pids == [] 
+
+  
+- name: Copying the python script
```
Run the script directly from its original location instead of making a copy to execute.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,59 @@
+# Test code for the pid module
+# Copyright 2018, Saranya sridharan
```
When including a copyright notice, also include the license:

```yaml
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

mattclay commented on test/integration/targets/pids/aliases:
```diff
@@ -0,0 +1,3 @@
+shippable/posix/group3
+skip/freebsd
+skip/osx
```
Why are these platforms skipped? Is `pidof` not available as a package on these systems?

saranyasridharan replied:
Yes. They are not in that platforms.

mattclay replied:
I just checked and `pidof` is available on macOS via `brew` and FreeBSD via `pkg`. Here's an example of how to install a package with `brew` in your test:

https://github.com/ansible/ansible/blob/f2f743aef039e4f7241b8d8e186baa11cbe625c9/test/integration/targets/archive/tasks/main.yml#L43-L60

So you should be able to include both those platforms in your test.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,73 @@
+      homebrew:
+        name: pidof
+        state: present
+        update_homebrew: no
```
Missing `become` as in this example:

https://github.com/ansible/ansible/blob/f2f743aef039e4f7241b8d8e186baa11cbe625c9/test/integration/targets/archive/tasks/main.yml#L43-L60

saranyasridharan commented:
@mattclay I am getting few errors , In osx , sample.py is not found and in freebsd , I am getting valueerror for empty results. Can you please help me with this?

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,79 @@
+    - emptypids.pids == [] 
+  when: ansible_os_family != 'FreeBSD'   
+  
+- name: "Changing file permission"
```
Make the file executable when you commit it to your branch instead of making it executable here.

mattclay commented on test/integration/targets/pids/aliases:
```diff
@@ -0,0 +1 @@
+shippable/posix/group3
```
Include a single newline at the end of the file.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,79 @@
+  pids: 
+    name: "blahblah"
+  register: emptypids
+  when: ansible_os_family != 'FreeBSD'
```
Remove this condition and fix the whitespace handling in the module.

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,64 @@
+def get_pid(name, module):
+    command = module.get_bin_path('pidof', True)
+    rc, processid, stderr = module.run_command([command, name])
+    return [int(pid) for pid in processid.split(" ") if processid]
```
Try this instead, to make sure that extra whitespace is handled correctly:

```python
rc, stdout, stderr = module.run_command([command, name])
return [int(pid) for pid in stdout.split()]
```

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,79 @@
+    that:
+    - emptypids is not changed
+    - emptypids.pids == [] 
+  when: ansible_os_family != 'FreeBSD'   
```
Remove this condition and fix the whitespace handling in the module.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,79 @@
+  async: 100
+  poll: 5
+  
+- name: "Getting the process IDs (PIDs) of python module through linux command"
```
Update the name to reflect the use of `checkobtainedpid`.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,79 @@
+  register: pidsobtained
+  failed_when: no
+      
+- name: "Checking the process IDs (PIDs) of python module" 
```
Update the name to reflect the use of `checkobtainedpid`.

saranyasridharan commented:
@mattclay thanks for the suggestions and helping me out. I have done the above mentioned changes.

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,64 @@
+def main():
+    module = AnsibleModule(
+        argument_spec={
+            "name": {"required": True, "type": "str"}}
```
Please move the last closing `}` into a new line (at same indentation level as `argument_spec`)

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,64 @@
+
+- name: Printing the process IDs obtained
+  debug:
+    msg:"PIDS of python:{{pids_of_python.pids|join(',')}}"
```
Please add some spacing, i.e.: `msg: "PIDs of python: {{ pids_of_python.pids | join(',') }}"`

felixfontein commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,64 @@
+
+RETURN = '''
+pids:
+  description: It return process IDs of the given process
```
You could drop "It return", i.e. change the description to "Process IDs of the given process"

felixfontein commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,85 @@
+    state: present
+  when: ansible_distribution == 'FreeBSD'
+  
+- name: Install  PidOf in (OSX)
```
Please change this message to `Installing PidOf in OSX`

felixfontein commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,85 @@
+  register: emptypids
+  
+   
+- name: "Verify the Process IDs (PIDs) returned is empty"
```
Do you mean: `Verify that the list of process IDs returned is empty`?

felixfontein commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,85 @@
+    name: "blahblah"
+  register: emptypids
+  
+   
```
Here (and below): please use only one newline between tasks, if there isn't a good reason for having more. In this place at least, it doesn't make sense that the above task is separated by its predecessors by only one newline and then the next task (which belongs to the above task) is suddenly separated by two newlines.

felixfontein commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,85 @@
+- name: "Verify the Process IDs (PIDs) returned is equal to the actual process IDs (PIDs) obtained"
+  assert:
+    that:
+    - "pids.pids | join(' ') == pidsobtained.stdout"
```
Don't you also want to assert that there is at least one PID?

saranyasridharan replied:
Is it needed @felixfontein ? Isn't the above condition check our requirement ?

felixfontein replied:
Wasn't the original motivation to have one test of the module where it returns a nontrivial result? If that is the case (otherwise: why go through all that trouble and install `setproctitle`) you really need this to make sure this actually happens. That this fails indicates that your test is **not** doing what you want it to do, so instead of removing this part of the test, you should fix it.

felixfontein replied:
I don't think your latest commit will fix this. The problem is that `checkobtainedpid` seems to be no longer running when the `pids` module and the `pidof` command are run.

saranyasridharan replied:
That is what I wonder. It is running in async, so after starting the task , it should come to next tasks. I am thinking what is missed here.

felixfontein replied:
Have you actually tried to run this locally (on a *nix system): execute `sample.py` in one shell, and running `pidof checkobtainedpid` in another shell? Did this actually return something?

Also note that your PR includes `sample.py` *twice*. I guess you only need one of the two copies.

saranyasridharan replied:
I have used some other variable{{ ansible_facts.env['PWD'] }} to get pwd . There was some issue with setproctitle installation earlier, I was able to get the pid of sample.py in the defined name after installating gcc and python-devel in my environment. But here it still fails

mattclay replied:
This comparison is probably failing because of extra spaces in the output.

> find process

```
04:19     "stdout_lines": [
04:19         "6192 "
04:19     ]
```

> Checking the process IDs (PIDs) of checkobtainedpid module

```
04:19     "pids": [
04:19         6192
04:19     ]
```

Remove the extra spaces and the comparison should work.

felixfontein reviewed (changes requested):
Some minor details

felixfontein commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,82 @@
+- name: Executing the python script
+  command: ./sample.py
+  args:
+    chdir: /var/root/ansible/test/integration/targets/pids/tasks
```
Can't you use `{{ role_path }}` to make this OS-independent? CC @mattclay

Right now, if the files are ever moved somewhere else, this will break.

mattclay replied:
@saranyasridharan Use of `chdir` shouldn't be needed. Why was this added?

saranyasridharan replied:
Without specifying path, I was not able to execute that. 'No file found error' occured. So I added this.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,82 @@
+    - emptypids.pids == [] 
+   
+- name: Executing the python script
+  command: "{{ ansible_facts.env['PWD'] }}/sample.py"
```
Include only a single copy of the script in: `test/integration/targets/pids/files/sample.py`

Reference it in the playbook using:

```yaml
command: "{{ role_path }}/files/sample.py"
```

mattclay replied:
Correction, `role_path`, not `role_dir`.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -61,7 +60,7 @@
    
 - name: Executing the python script
-  command: "{{ role_dir }}/files/sample.py"
+  command: "{{ ansible_facts.env['PWD'] }}/files/sample.py"
```
Sorry, my original comment was wrong (I corrected it later). It should have been `role_path`, not `role_dir`.

saranyasridharan commented:
@mattclay Do I need to set that role_path variable ?

mattclay commented:
No, it's set automatically when running an Ansible role, which is what the integration test is.

saranyasridharan commented:
@felixfontein @mattclay  I am trying to install gcc & python-devel package in freebsd and osx, But python-devel is not being installed in freebsd, I am not sure whether it will be available by default in the freebsd & osx. How do I resolve that ?

mattclay commented:
@saranyasridharan Why do you need to install gcc and python-devel?

mattclay commented:
I just realized there's an easier and more reliable way to get a custom process name for this test than using `setproctitle`. Locate the `sleep` binary on the system and make a copy using a randomly generated name in the temporary directory for the test (found using the `output_dir` variable). Then execute that copy of the sleep command.

Something like this:

```yaml
- name: pick a random process name
  command: echo "{{ 99999999 | random }}"
  register: random_name

- name: find the `sleep` binary
  command: command -v sleep
  register: find_sleep

- name: copy `sleep` binary
  copy:
    src: "{{ find_sleep.stdout }}"
    dest: "{{ output_dir }}/{{ random_name.stdout }}"
    mode: "u+rwx"

- name: run copy of `sleep` binary
  command: "{{ output_dir }}/{{ random_name.stdout }} 10"
  async: 10
  poll: 0

- name: find process
  command: pidof "{{ random_name.stdout }}"
  register: find_process

- name: show what we found
  debug: var=find_process
```

saranyasridharan commented:
@mattclay  , I have tried like you suggested, command -v sleep was working in it, but at end pidof returned empty result, same like earlier approach.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,83 @@
+  register: pids
+
+- name: show what we found
+  debug: var=find_process
```
This variable isn't defined anywhere.

felixfontein commented:
Why did you throw away the new testcase suggested by @mattclay, instead of simply fixing the wrong variable name?

saranyasridharan commented:
@mattclay I have tried like this , obtained pidof of 'pidof' command itself and verified it is not empty value. Is it enough for this testcase ? Even if I think of comparing it with the pid obtained from the linux command 'pidof pidof' It will differ as it is new command.

saranyasridharan commented:
> Why did you throw away the new testcase suggested by @mattclay, instead of simply fixing the wrong variable name?

I have tried that. It does not work for me. I was getting error while executing this command "command 
-v sleep" to get the sleep binary in most of the os combination ( freebsd, osx,linux ubuntu) . So I have tried this one instead. @felixfontein

felixfontein commented:
From what I see, `command -v sleep` was successful almost everywhere but for CentOS 6, OpenSUSE 42.3, and Ubuntu -- but there, the problem was not that sleep couldn't be found, but that `command` itself wasn't found. So instead of just throwing away the whole test, why not fix this specific problem instead? You could for example also try `which sleep`, and use the one of them which works (or either one if both work).

felixfontein commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -78,5 +78,5 @@
     that:
     - "pids.pids | length > 0"
-    - "pids.pids | join(' ')== find_process"
+    - "pids.pids | join(' ')== find_process.stdout"
```
There's an extra space at the end of `find_process.stdout` for some of the tests. You need to trim that off, for example using jinja2's [trim filter](http://jinja.pocoo.org/docs/2.10/templates/#trim): `"pids.pids | join(' ')== find_process.stdout | trim"`

felixfontein reviewed (approved):
I think it's good now! One last thing you should do is reduce the number of commits (by squashing them). The current number (120) is way too big.

saranyasridharan commented:
@felixfontein I have squashed and commit count is reduced now. Thank you

saranyasridharan commented:
@sivel @mattclay I have addressed your review comments. thanks so much

felixfontein commented:
Could you please change the commit message for the remaining commit from "Reducing the commit count" to something more sensible, for example "Adding custom module to find pid of the process" or "Add pids module"?

saranyasridharan commented:
> Could you please change the commit message for the remaining commit from "Reducing the commit count" to something more sensible, for example "Adding custom module to find pid of the process" or "Add pids module"?

Sure. It's done. Thanks

felixfontein reviewed (approved):
shipit

felixfontein commented:
@mattclay @sivel can you take a look whether the changes you requested have been addressed?

saranyasridharan commented:
@mattclay @sivel Is there any update on review changes approval?

saranyasridharan commented:
@mattclay @sivel Can you please verify the changes requested and accept it ?

mattclay reviewed (changes requested):
I just noticed the [comment](https://github.com/ansible/ansible/pull/43807#pullrequestreview-144651582) from @jamescassell about substring matches.

I tested `pidof` of macOS (installed via brew) and Ubuntu 18.04 and I'm seeing different behavior. On macOS substrings are matched, but on Ubuntu they are not.

Although the underlying `pidof` tool behaves differently on different platforms, I don't think that is desirable for this module. It doesn't look like `pidof` on macOS has an option for exact matching, so perhaps switching to another method of finding the process is needed. I recommend taking a look at `pgrep` to start with.

mattclay commented:
@saranyasridharan You may also want to take a look at https://pypi.org/project/psutil/ as well.

saranyasridharan committed:
Changing pidof to pgrep

saranyasridharan committed:
Correcting space issue

saranyasridharan committed:
Changing pgrep option

saranyasridharan committed:
Adding custom module to get PID for the given process name

saranyasridharan committed:
Adding custom module to get PID for the given process name

saranyasridharan committed:
Adding Custom module to find PID of the process name

saranyasridharan commented:
@mattclay I have used pgrep module, I guess it will solve that different output with different OS issue. Can you please review it ?

bcoca commented:
https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/service.py#L257 <= we use similar function for service modules and it currently deals with most cross platform issues we've found

mattclay commented:
@saranyasridharan Thank you for making those changes. I did some further testing and discovered that `pgrep -x` also has issues. On Ubuntu 18.04 it truncates process names before matching:

```shell
$ ps x | grep [g]nome-terminal
 4571 ?        Ssl    0:01 /usr/lib/gnome-terminal/gnome-terminal-server
$ pgrep -x gnome-terminal-server
$ pgrep -x gnome-terminal-
4571
```

Based on the suggestion from @bcoca I tried to use [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/ps.html) options to `ps` instead:

```
ps -A -o pid -o comm | grep [g]nome-terminal
 4571 gnome-terminal-
```

Next I tried using https://pypi.org/project/psutil/ with the following `findproc.py` script:

```python
import sys
import psutil

names = set(sys.argv[1:])

for proc in psutil.process_iter():
    if proc.name() in names:
        print(proc)
```

```shell
$ python findproc.py gnome-terminal-server
psutil.Process(pid=4571, name='gnome-terminal-server', started='2018-12-08 02:26:43')
```

It appears as though `psutil` may be the best cross platform option available.

I recommend that you include a long process name with spaces as part of your integration test, as that would have uncovered this issue.

saranyasridharan committed:
trying out psutil module

saranyasridharan committed:
adding script file

saranyasridharan committed:
adding pip install psutil

saranyasridharan committed:
correcting print errors

saranyasridharan committed:
correcting errors

saranyasridharan committed:
correcting errors

saranyasridharan committed:
correcting errors

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,63 @@
+
+from ansible.module_utils.basic import AnsibleModule
+import sys
+import psutil
```
Imports that are not part of the Python standard library need to be wrapped in a try/except block.

```suggestion
try:
    import psutil
    HAS_PSUTIL = True
else:
    HAS_PSUTIL = False
```

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,63 @@
+            "name": {"required": True, "type": "str"}
+        }
+    )
+    name = module.params["name"]
```
Check `HAS_PSUTIL` here:

```suggestion
    if not HAS_PSUTIL:
        module.fail_json(msg="Missing required 'psutil' python module. Try installing it with: pip install psutil")
    name = module.params["name"]
```

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,63 @@
+short_description: "Retrieves process IDs list if the process is running otherwise return empty list"
+author:
+  - Saranya Sridharan (@saranyasridharan)
+options:
```
```suggestion
requirements:
  - psutil
options:
```

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,51 @@
+
+- name: "Finding the process ID(PIDs)"
+  command: python {{ playbook_dir }}/targets/pids/tasks/obtainpid.py "{{ random_name.stdout }}"
+  register: find_process
```
This is effectively testing the same code as the module uses to get the pid. It might be a good idea to start the background sleep process using a bash or python script that writes the child pid to a file in the output_dir instead. That way we can check the module results against that instead.

mattclay commented on test/integration/targets/pids/tasks/obtainpid.py:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/python
```
```suggestion
#!/usr/bin/env python
```

saranyasridharan committed:
Committing new changes

saranyasridharan committed:
Committing new changes

saranyasridharan committed:
Correcting errors

saranyasridharan committed:
Correcting syntax error

saranyasridharan commented:
Thanks @mattclay for the help. I wonder why I am getting the module fail message, when I am installing psutil in test script before running any test.  Can you please help me with this ?

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,74 @@
+try:
+    import psutil
+    HAS_PSUTIL = True
+except Exception:
```
Use the following try/except, which is fairly standard for Ansible modules:

```python
try:
    import psutil
    HAS_PSUTIL = True
except ImportError:
    HAS_PSUTIL = False
```

This will also fix the test errors, as the existing code sets `HAS_PSUTIL` to `True` and then `False` when the import succeeds, but leaves it undefined when the import fails.

saranyasridharan committed:
Correcting import error

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

mattclay commented:
@saranyasridharan Did you know you can run the tests locally?

If you have Docker available, you can run the tests from your working copy:

```shell
source hacking/env-setup
ansible-test sanity pids test/integration/targets/pids/ --docker -v
ansible-test integration pids --docker ubuntu1604 -v
```

If you don't want to use Docker you can still run the tests locally.

For example, to run using tox, replace `--docker` with `--tox --python 2.7`.
If you're using your own venv, you can use `--requirements` instead.

saranyasridharan committed:
Correcting the errors

saranyasridharan committed:
Correcting the errors

mattclay commented:
Yes, you'll need `tox` installed to use the `--tox` option. If you don't want to use it, you can set up a virtual environment and run the tests there instead. For example, if your default Python is 2.7:

```shell
virtualenv ~/ansible-venv
. ~/ansible-venv/bin/activate
source hacking/env-setup
ansible-test sanity pids test/integration/targets/pids/ --requirements --python 2.7 -v
ansible-test integration pids --requirements -v
```

saranyasridharan committed:
Correcting the errors

saranyasridharan commented:
> Yes, you'll need `tox` installed to use the `--tox` option. If you don't want to use it, you can set up a virtual environment and run the tests there instead. For example, if your default Python is 2.7:
> 
> ```shell
> virtualenv ~/ansible-venv
> . ~/ansible-venv/bin/activate
> source hacking/env-setup
> ansible-test sanity pids test/integration/targets/pids/ --requirements --python 2.7 -v
> ansible-test integration pids --requirements -v
> ```

Okay I will do that . I could see almost all the test cases are passing now. Is it okay now ? or  Can I modify this random number to some long name , same way copy the sleep binary and all to check the name length case ??

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,52 @@
+    - emptypids.pids == [] 
+
+- name: "Picking a random process name"
+  command: echo "sample-name"
```
Try a long, random process name. Perhaps something like this:

```suggestion
  command: echo "some-long-random-process-name-{{ 99999999 | random }}"
```

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,52 @@
+    mode: "u+rwx"
+
+- name: "Running the copy of 'sleep' binary"
+  command: "sh {{ ansible_facts.env['PWD'] }}/targets/pids/tasks/obtainpid.sh {{ output_dir }}/{{ random_name.stdout }} {{ ansible_facts.env['PWD'] }}/targets/pids/tasks/"
```
```suggestion
  command: "{{ role_path }}/files/obtainpid.sh '{{ random_name.stdout }}' '{{ output_dir }}/obtainpid.txt'"
```

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,52 @@
+  register: pids
+    
+- name: "Reading pid from the file"
+  command: cat {{ ansible_facts.env['PWD'] }}/targets/pids/tasks/obtainpid.txt
```
```suggestion
  - slurp:
      src: "{{ output_dir }}/obtainpid.txt"
```

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,52 @@
+- name: "Verify that the Process IDs (PIDs) returned is not empty and also equal to the PIDs obtained in console"
+  assert:
+    that:
+    - "pids.pids | join(' ')  == newpid.stdout | trim"
```
```suggestion
    - "pids.pids | join(' ')  == newpid.content | b64decode | trim"
```

mattclay commented on test/integration/targets/pids/tasks/obtainpid.sh:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+$1 100 & 
+echo $! > "$2/obtainpid.txt"
```
```suggestion
echo $! > "$2"
```

mattclay reviewed (changes requested):
Make sure `test/integration/targets/pids/tasks/obtainpid.sh` is executable and remove `test/integration/targets/pids/tasks/obtainpid.txt` from the PR.

I suggested some changes to store `obtainedpid.txt` in the output dir instead of the working tree, and also moving `obtainedpid.sh` into the `files` directory.

saranyasridharan committed:
Addressing the changes

saranyasridharan committed:
Addressing the changes

saranyasridharan committed:
Addressing the changes

Commit:
Merge branch 'pid' of https://github.com/saranyasridharan/ansible into devel

Commit:
correcting errors

Commit:
correcting errors

saranyasridharan committed:
Correcting errors

saranyasridharan committed:
Correcting errors

saranyasridharan committed:
Correcting errors

saranyasridharan committed:
Correcting errors

saranyasridharan committed:
Correcting errors

saranyasridharan committed:
Correcting errors

saranyasridharan committed:
Correcting errors

saranyasridharan committed:
Correcting errors

saranyasridharan commented:
@mattclay I am almost done with the long process name change in the test case. I am only getting error for freebsd system. Can you please help me with this ?

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,72 @@
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+import sys
```
The `sys` import is unused.

```suggestion
```

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,72 @@
+    HAS_PSUTIL = False
+
+
+def get_pid(name, module):
```
The `module` parameter is unused.

```suggestion
def get_pid(name):
```

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,72 @@
+
+
+def get_pid(name, module):
+    return [int(p.info['pid']) for p in psutil.process_iter(attrs=['pid', 'name']) if name in p.info['name']]
```
Match name exactly.

```suggestion
    return [int(p.info['pid']) for p in psutil.process_iter(attrs=['pid', 'name']) if name == p.info['name']]
```

This will also fix the CI failure, since it's failing when `p.info['name']` is `None`.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,53 @@
+  assert:
+    that:
+    - "pids.pids | join(' ')  == newpid.content | b64decode | trim"
+    - "pids.pids | length > 0"
```
Add a test using a substring of `random_name.stdout` to make sure only exact matches work.

saranyasridharan commented:
@mattclay. Thank you so much for the review. I have addressed the above mentioned points. Please let me know if anything else is needed.

saranyasridharan commented:
@mattclay  @felixfontein I have stashed commit count and pulled changes from devel branch in ansible.git, should I rebase also ?

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,72 @@
+
+
+def get_pid(name):
+    return [int(p.info['pid']) for p in psutil.process_iter(attrs=['pid', 'name']) if name == p.info['name']]
```
The cast to `int` shouldn't be needed, as `p.info['pid']` should already be an int.

```suggestion
    return [p.info['pid'] for p in psutil.process_iter(attrs=['pid', 'name']) if name == p.info['name']]
```

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,72 @@
+    module = AnsibleModule(
+        argument_spec={
+            "name": {"required": True, "type": "str"}
+        }
```
Although it's cosmetic, nearly all of our modules use the following syntax since it uses fewer quotes:

```python
        argument_spec=dict(
            name=dict(type='str', required=True),
        ),
```

It also includes trailing commas to make adding additional lines later cleaner.

mattclay commented on lib/ansible/modules/system/pids.py:
```diff
@@ -0,0 +1,72 @@
+        argument_spec={
+            "name": {"required": True, "type": "str"}
+        }
+    )
```
Since the module makes no changes, it's safe to claim support for check mode:

```suggestion
        supports_check_mode=True,
    )
```

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,59 @@
+- name: "copying 'sleep' binary"
+  copy:
+    src: "{{ find_sleep.stdout }}"
+    dest: "{{ role_path }}/tasks/{{ random_name.stdout }}"
```
Use the temporary output dir for storing temporary files:

```suggestion
    dest: "{{ output_dir }}/{{ random_name.stdout }}"
```

mattclay commented on test/integration/targets/pids/tasks/files/obtainpid.sh:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
```
This should be in `test/integration/targets/pids/files/obtainpid.sh` instead, since it's convention for Ansible roles to have a `files` directory at the same level as the `tasks` directory:

https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#role-directory-structure

mattclay commented on test/integration/targets/pids/tasks/obtainpid.txt:
```diff
@@ -0,0 +1 @@
+7476
```
This file shouldn't be checked in.

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,59 @@
+    mode: "777"
+
+- name: "Running the copy of 'sleep' binary"
+  command: "sh {{ role_path }}/tasks/files/obtainpid.sh '{{ role_path }}/tasks/{{ random_name.stdout }}' '{{ role_path }}/tasks/obtainpid.txt'"
```
```suggestion
  command: "sh {{ role_path }}/files/obtainpid.sh '{{ output_dir }}/{{ random_name.stdout }}' '{{ output_dir }}/obtainpid.txt'"
```

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,59 @@
+  
+- name: "Reading pid from the file"
+  slurp:
+    src:  "{{ role_path }}/tasks/obtainpid.txt"
```
```suggestion
    src:  "{{ output_dir }}/obtainpid.txt"
```

mattclay commented on test/integration/targets/pids/tasks/main.yml:
```diff
@@ -0,0 +1,59 @@
+    name: "{{ random_name.stdout }}"
+  register: pids
+
+- name: "new testcase"
```
Use a more descriptive task name:

```suggestion
- name: "Checking that exact non-substring matches are required"
```

mattclay reviewed (changes requested):
The module and tests are looking much better. There are some small things that should be cleaned up before merging.

saranyasridharan commented:
@mattclay Thanks for the review comments. I have addressed the same. Do I need to do rebase the code to remove the merge commit, because I am seeing the needed_rebase label in the PR? is anything else pending from my side after this ?

mattclay commented:
@saranyasridharan Yes, you'll need to rebase to remove the merge commit, which shows up as `Merge branch 'devel' of git://github.com/ansible/ansible into pid` (commit b31ff6dc2cc71b860bc0f92c410c39f5b683e149).

saranyasridharan committed:
Revert "Adding custom module to find PID of the process"

This reverts commit d5667d5f4ac859fc945c0dc6b26f6eb9c2eb7fdc.

saranyasridharan committed:
Adding custom module to find PID of the process

saranyasridharan committed:
Update pids.py

Addressing the review comments.

saranyasridharan committed:
Update pids.py

saranyasridharan committed:
Adding custom module to get PID of the process

saranyasridharan commented:
@mattclay @felixfontein I was having some problem with rebase, I have recreated this branch, but it didn't help. Now this pull request is closed , Can I reopen It ? or should I create new one ? :( Finally before rebase all test were passed and module was working as expected.

mattclay commented:
@saranyasridharan You can re-open this one and fix the rebasing issue, or if it's easier, create a  new one. If you create a new one, add a comment to this one referring to the new PR.

saranyasridharan commented:
@mattclay Thanks for the suggestion. I have created new PR. 
@mattclay & @felixfontein Please find the link to new PR .
